pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -244,74 +250,123 @@ public SuccessResponse deleteAllSegments(
         "All segments of table " + TableNameBuilder.forType(tableType).tableNameWithType(tableName) + " deleted");
   }
 
-  @POST
-  @ManagedAsync
-  @Produces(MediaType.APPLICATION_JSON)
-  @Consumes(MediaType.MULTIPART_FORM_DATA)
-  @Path("/segments")
-  @ApiOperation(value = "Upload a segment", notes = "Upload a segment as binary")
-  public void uploadSegmentAsMultiPart(FormDataMultiPart multiPart,
-      @ApiParam(value = "Whether to enable parallel push protection") @DefaultValue("false") @QueryParam(FileUploadDownloadClient.QueryParameters.ENABLE_PARALLEL_PUSH_PROTECTION) boolean enableParallelPushProtection,
-      @Context HttpHeaders headers, @Context Request request, @Suspended final AsyncResponse asyncResponse) {
+  private SuccessResponse uploadSegment(FormDataMultiPart multiPart, boolean enableParallelPushProtection, HttpHeaders headers, Request request) throws Exception {
+    // Get upload type
+    String uploadTypeStr = headers.getHeaderString(FileUploadDownloadClient.CustomHeaders.UPLOAD_TYPE);
+    FileUploadDownloadClient.FileUploadType uploadType = getUploadType(uploadTypeStr);
+
+    // Get crypter class
+    String crypterClassHeader = headers.getHeaderString(FileUploadDownloadClient.CustomHeaders.CRYPTER);
+
+    // Get Download URI
+    String downloadURI = headers.getHeaderString(FileUploadDownloadClient.CustomHeaders.DOWNLOAD_URI);
+
+    File tempEncryptedFile = null;
+    File tempDecryptedFile = null;
+    File tempSegmentDir = null;
     try {
-      asyncResponse.resume(uploadSegmentInternal(multiPart, null, enableParallelPushProtection, headers, request, false));
-    } catch (Throwable t) {
-      asyncResponse.resume(t);
+      FileUploadPathProvider provider = new FileUploadPathProvider(_controllerConf);
+      String tempFileName = "tmp-" + System.nanoTime();
+      tempEncryptedFile = new File(provider.getFileUploadTmpDir(), tempFileName + "_encrypted");
+      tempDecryptedFile = new File(provider.getFileUploadTmpDir(), tempFileName);
+      tempSegmentDir = new File(provider.getTmpUntarredPath(), tempFileName);
+
+      // TODO: Change when metadata upload added
+      String metadataProviderClass = DefaultMetadataProvider.class.getName();
+
+      Object encryptedObject;
+      switch (uploadType) {
+        case URI:
+          if (downloadURI == null || downloadURI.isEmpty()) {
+            throw new ControllerApplicationException(LOGGER, "Failed to get downloadURI, needed for URI upload",
+                Response.Status.BAD_REQUEST);
+          }
+          if (crypterClassHeader == null) {
+            crypterClassHeader = DefaultPinotCrypter.class.getName();
+          }
+          encryptedObject = tempEncryptedFile;
+          break;
+        case TAR:
+          // In the case of TAR, we will use local file
+          if (crypterClassHeader == null) {
+            crypterClassHeader = MultipartPinotCrypter.class.getName();
+          }
+          downloadURI = "multipart://dummy";
```

Comment:
We want to plug all upload methods into a SegmentFetcher as we discussed during the design, so that's why I've created a multipart one that is a noop.